### ISSUE TITLE
How should I give max_val in QuantReLU

### ISSUE THREAD

**Issue by ziqi-zhang** [Xilinx/brevitas#71]
Hi,
Thanks for sharing the awesome code. I have a question about the max_val parameter in the QuantReLU. In the LeNet example and examples/imagenet_classification/models/common.py, it is set to 6. I was wondering if I want to construct my own net on image, or change bit_width from 8 to 4, do I need to change this parameter?

Thanks,

**Comment by Giuseppe5** [comment]
The max_val in QuantReLU (together with min_val in other activations such as QuantHardTanh) is used to initialize the scale factor which then will be updated according to the strategy chosen through scaling_impl_type (or kept constant if scaling_impl_type = 'CONST').

As a general rule of thumb, you want to reduce the max_val value as you reduce the bit_width, otherwise you will end up having a very coarse representation of a very large range of values, at least at the beginning of the training process, and then it may be difficult to converge to an optimal value.

**Comment by ziqi-zhang** [comment]
@giuliogamba thanks for your reply!

**Comment by volcacius** [comment]
To expand on @Giuseppe5's comment, in general max_val is a hyperparameter. Something that you can do, if you don't know how to initialize it, is to first to train briefly with scaling impl type == STATS with stats op == MAX, and then set scaling impl == PARAMETER. What this does is compute statistics at training time, while collecting an exponential moving average for inference (same principle as batch norm), and then use the ema to initialize the parameter. Brevitas takes care of the conversion (you can see a small example [here](https://github.com/Xilinx/brevitas/blob/master/test/test_act_scaling.py))

Alessandro
